[#64] Add action scheduling and make all needed data persistent

ponchofiesta opened the pull request.

pbroman commented on kitodo-mediaserver-core/src/main/java/org/kitodo/mediaserver/core/actions/WorkLockAction.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    public Object perform(Work work, Map<String, String> parameter) throws Exception {
+        Boolean enabled = Boolean.parseBoolean(parameter.get("enabled"));
+        work.setEnabled(enabled);
```
This will break with an NPE if the parameter map is null. Also it will always disable the work if the parameter "enabled" is not set in the map, with no questions asked. Suggestion: use the method checkForRequiredParameter in MedaServerUtils to check that the param is set and the map not null (just added the null check, please fetch...)

ponchofiesta commented:
TODO for me:
- check for NPE
- function/procedure in ActionService.request
- Tests

ponchofiesta commented:
I have added some null checks and tests. I didn't add ActionService.requestAndPersist() because its usage is to rare. The WorkLockAction is the only one which would need a better solution but as it is not the targeted use case I think it's OK. What do you think?

pbroman commented on kitodo-mediaserver-core/src/main/java/org/kitodo/mediaserver/core/services/ActionService.java:
```diff
@@ -0,0 +1,183 @@
+     * @return the result of the Action
+     * @throws Exception on Action errors
+     */
+    public Object performRequested(Work work, String actionName, Map<String, String> parameter) throws Exception {
```
This method can only be performed, when you already know the action in question. For the case, that one process request an action and another looks for actions to perform at some later time, that will not be the case. Then you would have a DB-query getting ActionData objects to perform and would expect a moethod performRequested(AcionData actionData). Perhaps we should offer both methods, the now implemented one calling the second after getting the ActionData object.

pbroman commented on kitodo-mediaserver-core/src/main/java/org/kitodo/mediaserver/core/services/ActionService.java:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    /**
+     * Finds the unfinished action (not running and running).
```
The difference between unfinished and running action is a bit confusing, to me, unfinished suggests it is already running, and "not running and running" doesn't really explain it... Please write a few more words in the javadoc. (Suggestion: Finds a requested action that hasn't already been performed, whether it is currently running or not.) Maybe getUnperformedAction would be better...

pbroman commented on kitodo-mediaserver-core/src/main/resources/schema.sql:
```diff
@@ -33,3 +33,22 @@ CREATE TABLE `user` (
+  FOREIGN KEY (`work_id`) REFERENCES `work`(`id`)
+);
+
+CREATE TABLE `parameter` (
```
Maybe we should call the table action_parameter

ponchofiesta replied:
The default would be "action_data_parameter" because it is part of ActionData. And if I change it in ActionData, the column names (key and value) will change too (action_paramater_key and action_paramater respectively action_data_parameter_key and action_data_parameter).

pbroman commented on kitodo-mediaserver-core/src/test/java/org/kitodo/mediaserver/core/services/ActionServiceTest.java:
```diff
@@ -0,0 +1,284 @@
+    }
+
+    @Test
+    public void requestAction() throws ActionServiceException {
```
It is not very clear what is actually being tested. I think it generally would be better, if the test method was named after what is being tested / what is expected rather than after which method is called or what the input is. However, we never really discussed this, so it would probably be good to confront the mattern in a meeting.

pbroman commented:
Basically it's looking good, I've added a few comments though, and since this is a component of central importance to the extendability of the mediaserver, we should put a lot of effort into it to make it work well.
Checkstyle is also not quite happy. ;)

ponchofiesta commented:
OK, next try. I've added `performRequested(ActionData)`. I've renamed Unifinished to Unperformed and added your comment. `requestAction` is still requestAction as I think this as exactly what happens there (and I don't know a more suitable name). But I also added further tests for this method and now for the first time for `perfomRequested`. Have a look. Maybe it helps to separate test cases.

There are checkstyle violations. But they ae in other files which are not modified in my commits. I'm currently working on a fix for the checkstyle maven checks. I would add those style fixes there.

pbroman commented:
Sieht gut aus!
